subsys/net/lib/lwm2m/lwm2m_obj_firmware.c

```diff
@@ -116,6 +117,10 @@ void lwm2m_firmware_set_update_state(uint8_t state)
 			update_state, state);
 	}
 
+	if (update_state_cb) {
+		update_state_cb(state);
+	}
+
 	update_state = state;
 	NOTIFY_OBSERVER(LWM2M_OBJECT_FIRMWARE_ID, 0, FIRMWARE_STATE_ID);
```

Comment:
I think that might be the reason post_write cb might now work - the FW Update object implementation shouldn't modify the state behind the engine's back, but rather use `lwm2m_engine_set_u8()` to update the state - that way all registered callbacks will be executed correctly for the resource.

Reply:
yes, exactly that is the point

Reply:
After thinking a bit about your comment I think it is better to use lwm2m_engine_set_u8() for upstate_state as well as for update_result and go with using pre/post_write callbacks. Do you agree on that?

Reply:
Yes, I totally agree.